core/logger/zap.go

```diff
@@ -5,31 +5,80 @@ import (
 	"fmt"
 	"io"
 	"os"
+	"path/filepath"
 
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
+	"gopkg.in/natefinch/lumberjack.v2"
 )
 
 var _ Logger = &zapLogger{}
 
 type zapLogger struct {
 	*zap.SugaredLogger
-	config     zap.Config
+	config     *Config
 	name       string
 	fields     []interface{}
 	callerSkip int
 }
 
-func newZapLogger(cfg zap.Config) (Logger, error) {
-	zl, err := cfg.Build()
-	if err != nil {
-		return nil, err
+func newZapLogger(cfg *Config) (Logger, error) {
+	var core zapcore.Core
+	if cfg.ToDisk {
+		core = zapcore.NewTee(newConsoleCore(cfg), newDiskCore(cfg))
+	} else {
+		core = newConsoleCore(cfg)
+	}
+	return &zapLogger{config: cfg, SugaredLogger: zap.New(core).Sugar()}, nil
+}
+
+func newDiskCore(cfg *Config) zapcore.Core {
+	var (
+		encoder = zapcore.NewJSONEncoder(makeEncoderConfig(cfg))
+		sink    = zapcore.AddSync(&lumberjack.Logger{
+			Filename:   filepath.ToSlash(filepath.Join(cfg.Dir, "log.jsonl")),
+			MaxSize:    cfg.DiskMaxSizeBeforeRotate,
+			MaxAge:     cfg.DiskMaxAgeBeforeDelete,
+			MaxBackups: cfg.DiskMaxBackupsBeforeDelete,
+			Compress:   true,
+		})
+		allLogLevels = zap.LevelEnablerFunc(func(lvl zapcore.Level) bool { return true })
+	)
+	return zapcore.NewCore(encoder, sink, allLogLevels)
+}
+
+func newConsoleCore(cfg *Config) zapcore.Core {
+	filteredLogLevels := zap.LevelEnablerFunc(func(lvl zapcore.Level) bool { return lvl < cfg.LogLevel })
```

Comment:
Does this race with `SetLogLevel`? Could use a `zap.AtomicLevel` instead, and pass `cfg.LogLevel.Enabled`.